### ISSUE TITLE
Add .env file support to COMPOSE_ENV_FILES

### ISSUE THREAD

**Issue by nickjj** [docker/compose#11122]
### Description

This is in relation to the new feature introduced in this PR: https://github.com/docker/compose/pull/11061

When `COMPOSE_ENV_FILES` is defined in an `.env` file it has no effect. This is a potential oversight / bug.

Here's a reproduceable example:

```yml
# docker-compose.yml
services:
  debian:
    image: "debian:bookworm-slim"
    restart: "${DOCKER_RESTART_POLICY:-unless-stopped}"
```

```sh
# .env
COMPOSE_ENV_FILES=.env,.env.local
DOCKER_RESTART_POLICY=yes
```

```sh
# .env.local
DOCKER_RESTART_POLICY=no
```

If you run `docker compose config` the value of restart is `yes` but it would be expected that it's `no` through the logic of:

- `docker compose` was called without explicitly defining any `--env-file` flags which means it will look for `.env` by default
- The new `COMPOSE_ENV_FILES` env var will be used which says to interpolate the `.env` file and then `.env.local`
  - This is the part that is failing the use case expectation
- `no` should be set

If you run `COMPOSE_ENV_FILES=.env,.env.local docker compose config` then everything works as expected but this requires altering your shell or creating a custom alias to run docker compose with that set. The original use case was to allow for running `docker compose` commands without modification.

**Comment by ndeloof** [comment]
I'm :-1 on this.
`COMPOSE_ENV_FILES` let you define the env files you want to include, as an alternative to a (potentially long) list of .env files to be included. From this point of view, this is an improvement.
With your proposed change, we are just back to defaulting to `.env`. If `COMPOSE_ENV_FILES` in this file doesn't include `.env` then resulting environment would just be inconsistent. And another env file might override this variable and include other files, what would then be the expected result ?
Basically, I'd say "this is going too far"

**Comment by nickjj** [comment]
As it currently stands (with `COMPOSE_ENV_FILES` as it is), we can do:

- `docker compose --env-file .env --env-file .env.custom build`
- `COMPOSE_ENV_FILES=.env,.env.custom docker compose build`

In both cases if you want `docker compose` to interpolate `.env.custom` you must modify how you use Docker Compose. You can no longer just call `docker compose`. You can do 1 of the following:

- Run either of the 2 options in the first set of bullets above
- Manually `export COMPOSE_ENV_FILES=.env,.env.custom` in your current shell
- Create an alias for `docker compose` but build out a lot of logic because you probably don't want this to apply to projects
- Modify your shell's rc file to auto-export `COMPOSE_ENV_FILES` but like the alias approach, this needs a lot of logic to work

In my opinion all 4 of the above cases are not a great developer experience since it involves every person who runs your project to customize how they call `docker compose` in one of the above ways.

This issue (and the original discussion) allows `COMPOSE_ENV_FILES` to be defined in `.env` and Docker Compose will look for that automatically. If it exists then it will interpolate all files in that list. The only extra step / functionality is this behavior. Everything else would work the same once `COMPOSE_ENV_FILES` is identified.

**Comment by nickjj** [comment]
> indeed, this is what environment variables are for: tweak your environment so the same command will behave in a distinct way.

Right but now imagine you have 30 developers who are working on an app.

Now to run the project all 30 developers need to run `export COMPOSE_ENV_FILES=.env,.env.custom` every time they want to start up the project. Also, assume 75% of the developers don't even know what that command does because they are application developers and not necessarily heavy terminal users.

Likewise, imagine you open source a project that's used by thousands of folks. Suddenly you can't just `docker compose up --build` the project anymore. You need to export an environment variable. Not just once, but every single time you open a terminal.

**Comment by ndeloof** [comment]
having multiple env file suggest you want to combine those to cover multiple scenarios. If the default `.env` one automagically include more, then .. why not have all those variables set in a single file ?

**Comment by nickjj** [comment]
It wouldn't be automagical. It would be dependent on you explicitly setting `COMPOSE_ENV_FILES` in the `.env` file. If you don't set that env var in `.env` then Docker Compose would only interpolate variables in `.env`.

There's a number of use cases where you couldn't or wouldn't want to set everything in a single `.env`. Maybe you have a separate `.env.local` file which isn't commit to version control and it's meant to be local env vars to 1 specific developer where as your `.env` file is shared across multiple developers and commit to version control. You may also have a separate `.env.secrets` file for anything not safe to commit.

In any of the above cases you could or would want Docker Compose to interpolate, `.env`, `.env.local` and `env.secrets`.

**Comment by nickjj** [comment]
I think one last thing to bring up are the existing `COMPOSE_*` env variables. By default Docker Compose will pick them up and reference them when running `docker compose` commands without needing to manually export or reference `--env-file`.

Besides covering the use case description to spawn this feature, from an intuitiveness perspective the expected outcome is `COMPOSE_ENV_FILES` works the same as the other `COMPOSE_*` variables.

**Comment by ndeloof** [comment]
what if .env defines `COMPOSE_ENV_FILES=x.env` and x.env also defines `COMPOSE_ENV_FILES=y.env`?
the resulting COMPOSE_ENV_FILES would then list files which where not actually used to load its own value 🤪

IMHO this really goes way beyond usability rules, and I'll veto such a feature

**Comment by nickjj** [comment]
Ok no problem.

You may want to consider reverting `COMPOSE_ENV_FILES` entirely because as it currently stands it's really misleading since every other `COMPOSE_*` env var gets picked up by `.env` by default. This will be the only exception. As is, it also doesn't help solve the use case of wanting to reference multiple env files without needing every developer / user to modify their local shell.

When you put both of those together, I'm not sure how much value there is to having `COMPOSE_ENV_FILES` when you could just pass in `--env-file` flags. In either case if you plan to use this feature as is you need to do something custom to make it work and adding an env var or flag are both custom things.

**Comment by rLoka** [comment]
Has there been any news to this? Official documentation is misleading, it does not work as described there: https://docs.docker.com/compose/environment-variables/envvars/#compose_env_files

I am getting this in case I want to load vars from other env specified (upon **up -d**).
```
WARN[0000] The "VAR1" variable is not set. Defaulting to a blank string. 
WARN[0000] The "VAR2" variable is not set. Defaulting to a blank string. 
...

```

This would be an example of additional **.env**:
```
COMPOSE_ENV_FILES=.env.additional
```

Content of **.env.additional**:
```
VAR1=foo
VAR2=bar
```

**Comment by RicardoM17** [comment]
> It wouldn't be automagical. It would be dependent on you explicitly setting `COMPOSE_ENV_FILES` in the `.env` file. If you don't set that env var in `.env` then Docker Compose would only interpolate variables in `.env`.
> 
> There's a number of use cases where you couldn't or wouldn't want to set everything in a single `.env`. Maybe you have a separate `.env.local` file which isn't commit to version control and it's meant to be local env vars to 1 specific developer where as your `.env` file is shared across multiple developers and commit to version control. You may also have a separate `.env.secrets` file for anything not safe to commit which is populated however you see fit (external secret store, etc.).
> 
> In any of the above cases you could or would want Docker Compose to interpolate, `.env`, `.env.local` and `env.secrets`.

I am exactly in this use case and it's extremely sad to see such an issue that was raised and then it was vetoed. Unfortunately this is not the first time that this has happened with `docker` and `compose` projects. Honestly it seems that the maintainers are more worried about setting arbitrary boundaries that then completely handicap some of the features. Coincidentally another example of this is also regarding docker `profiles`.

Even the author of the feature (@glours ) admitted that not adding support via the `.env` was just an oversight that then @nickjj was kind enough to correct.

Plenty of people that use this tool have different use cases and requirements, that is fine and expected. Yes you can't support every single use case but many of them are downright obvious and more than relevant, like in this scenario. I cannot understand why this project goes so out of their way to reject help from other users while at the same time making their experience worse.

Just my 2 cts.

**Comment by RicardoM17** [comment]
@ndeloof then how is one supposed to set the `.env` files to be used by `compose` in a way that is **scalable and reliable** and **most importantly is project in/dependent**? I see your concerns as to how this can get out of hand if we are defining the `.env` files to be used by `compose` within the `.env` file itself. But while that is a valid concern it does not erase the issue.

Having to manually export a variable every single time or having to add lengthy options to a command is completely impractical.

More "permanent" solutions like aliasing the command or adding such a variable export to `.bashrc` do not properly work when one has more than 1 `docker` project.

So if we can't have this variable defined in `.env` file, even though this an oversight from the initial implementation (see my comment above), can we have it somewhere else? Like for example in the `compose.yaml` file?

Thanks

**Comment by puco** [comment]
If this feature is not going to be implemented at least you can mention this in the documentation, which blankly states, that pre-defined environment docker compose env. variables can be set using the `.env` file (without exceptions) (https://docs.docker.com/compose/how-tos/environment-variables/envvars/). I tried using this until I stumbled across this GH issue to see that my approach to defining the compose file is futile.

**Comment by b-a0** [comment]
> If this feature is not going to be implemented at least you can mention this in the documentation, which blankly states, that pre-defined environment docker compose env. variables can be set using the `.env` file (without exceptions) (https://docs.docker.com/compose/how-tos/environment-variables/envvars/). I tried using this until I stumbled across this GH issue to see that my approach to defining the compose file is futile.

Unfortunately I ran into the same issue. Please update the docs (or bring functionality in line with the docs). Thanks in advance!

**Comment by sergiorcs82** [comment]
This is still an issue.

The stack i am working on is used by several developers who know little to nothing about Docker or the command line.

The standard dev doesn't much like to read or type.

Using a custom script is a slippery slope i would rather avoid.
It's happened before: it starts as something simple, but people keep requesting/adding tiny features to a point where we can no longer use standard commands directly, due to complexity.
The script then becomes a blackbox, and devs end up learning nothing about Docker and thinking that it is a can of worms.

By setting a rule about not using scripts, i myself was able to learn much more about Docker and Compose, as i needed to rely on standard features instead of hacks.

I stumbled upon `COMPOSE_ENV_FILES` in the official documentation.
This was perfect: i can split the `.env` file into something more intuitive and manageable. Users can keep using `docker compose up -d` without hassle.

It was disappointment i found this thread.

I hear the argument about `COMPOSE_ENV_FILES` being defined recursively.
This is not unheard of and there are algorithms to handle such situations. Off the top of my head:
- Before loading a file, check a list of loaded files. Skip file if it exists in list.
- Add file to list.
- Load file.
Or, if you want to be stricter/simpler: ignore `COMPOSE_ENV_FILES` if the current file isn't the main `.env`.
Even this would be better that what we already have.

I am all in favor of keepings things simple.
But i am also in favor of keeping things consistent: this is the only `COMPOSE_*` variable that doesn't work like the others.
And i am in favor of avoiding redundancy: as it is, `COMPOSE_ENV_FILES` adds little to what one can already do with `--env-file`.

Considering that this would not be a breaking change, i have trouble understanding why it can't be made work the way 17+ users have already requested.

**Comment by beechfuzz** [comment]
I too am wanting to use this `COMPOSE_ENV_FILES` feature as it's described in the docs, yet it's still not working as of today.

**Comment by narthur-asentry** [comment]
Just another vote for this feature. We use docker to deploy software to many different environments, and the people doing those deploys aren't docker experts. We currently have a shell script wrapper, but it has gotten messy and hard to understand, so we're trying to get rid of it. `COMPOSE_ENV_FILES` felt like the perfect way to solve the problem, because we could have an `.env.local` that would let us set `COMPOSE_PROFILES` to match the software deployed on that machine, and then our admins could just use `docker compose` without having to string together a bunch of arguments every time. 

And, as has been repeatedly mentioned, the docs make it seem like this is possible. But then we went to implement it, and discovered that it's not. Very frustrating.

However, the underlying chicken-and-egg problem seems real. So what about making a new environment variable, `COMPOSE_ADDITIONAL_ENV`, that explicitly means "additional env files to load, beyond the current/default one"? (And it could be set from `.env`.) Then there's no chicken-and-egg problem, and everybody trying to use this feature can accomplish what they're trying to accomplish.

**Comment by arcxyz** [comment]
@narthur-asentry 
Maybe a different angle that sidesteps the chicken-and-egg entirely: instead of trying to load *additional* env files, just swap what the **default `.env`** points to with a symlink.

The default `.env` in the working directory is special. Compose loads it first and automatically, *before* resolving profiles, and `COMPOSE_PROFILES` / `COMPOSE_FILE` set there are honored. That's exactly the variable that doesn't resolve when set via `COMPOSE_ENV_FILES`. So the fix is to never ask Compose to load a *different* file (the part that's broken), and instead repoint the one file it always reads:

```bash
# switch-env.sh <env>
ln -sf "overlays/$ENV/.env" .env
```

Each environment's file carries its own profiles:

```bash
# overlays/production/.env
COMPOSE_FILE=compose.yml:overlays/production/compose.override.yml
COMPOSE_PROFILES=web,api,worker,admin,monitoring,logging

# overlays/qa/.env
COMPOSE_PROFILES=web,api,worker,admin,monitoring,logging,devtools
```

After `./switch-env.sh production`, your admins just run plain `docker compose up -d`, with no `--profile`, no `--env-file`. The active profiles come from the symlinked file. We use this in production across multiple environments and it's been rock solid.

This is essentially your `.env.local` idea, except it *replaces* the default `.env` rather than layering on top of it. The trade-off: you don't get base + machine-overrides layering out of the box, since each target is one self-contained file. If you do need layering, the default `.env` can also set `COMPOSE_ENV_FILES` to chain more files, and *that* is honored because `.env` is read first (the same reason the symlink works).

Doesn't replace the feature request, but it might unblock you today.

**Comment by narthur-asentry** [comment]
@arcxyz this is interesting... but we do depend on `.env` file layering. Our installer installs a version-specific `.env` file, with a bunch of variables, and then we use a secondary/local file for machine-specific overrides. (And that's where we wanted to set `COMPOSE_PROFILES`.)

But you said two things that caught my attention:

> The default `.env` in the working directory is special. Compose loads it first and automatically, _before_ resolving profiles, and `COMPOSE_PROFILES` / `COMPOSE_FILE` set there are honored. That's exactly the variable that doesn't resolve when set via `COMPOSE_ENV_FILES`. 

That's news to me! `COMPOSE_PROFILES` isn't obeyed, in an env file referenced by `COMPOSE_ENV_FILES`? If so, that's another reason our idea couldn't have worked. 

> If you do need layering, the default `.env` can also set `COMPOSE_ENV_FILES` to chain more files, and _that_ is honored because `.env` is read first (the same reason the symlink works).

Isn't that exactly what this whole Issue is about - that that doesn't work? But you're saying it does?

**Comment by arcxyz** [comment]
Good catches, you're right on both and my memory had failed me. I was sure our project chained env files with `COMPOSE_ENV_FILES` set inside `.env`, but it doesn't: we just duplicate about half our variables across each environment's file, which is precisely the symptom of this not working.

The one practical upshot for your layering case: `COMPOSE_PROFILES` *is* honored when it lives in a file referenced by `COMPOSE_ENV_FILES`, so your version-specific file plus a local override can work as you wanted. The only catch is the one this issue is about: `COMPOSE_ENV_FILES` itself has to be set in the real environment, not in a file. Since your installer already touches the machine, dropping it once in `/etc/environment` or a systemd unit is a much smaller surface than the wrapper you're trying to kill.

**Comment by sergiorcs82** [comment]
I too went the way of the symlink - a simple, permanent one just so that i could control its name: `.env` is hidden by default, rendering it slightly less easy to use by operators of the stack.

Otherwise, i completely let go of the idea of using `COMPOSE_ENV_FILES`: i say this feature is intentionally broken/mangled into its current near-uselessness.

Instead, my symlinked `build.env` just always holds all env vars needed at build time, and i've leveraged `COMPOSE_FILE` to split my service definitions and `COMPOSE_PROFILES` to selectively load them.

I coupled this with a `runtime.yaml` that contains variables needed during runtime. These are loaded in a base compose file and perused by all services/containers.
This allows me to change some values without rebuilding them.

On top of this, the repo is split into two long-lived branches: one, production runtime; the other, an overlay where i place any development-related extras (e.g. build targets in docker images that layer extra package installs; adjustments to compose services, like setting said image targets; or extra vars or different values in that `build.env`).

I know this doesn't offer a solution to your problem, but hopefully it will give you ideas.
